internal/packager/fork_test.go

```diff
@@ -0,0 +1,97 @@
+package packager
+
+import (
+	"testing"
+
+	"github.com/docker/app/internal/types"
+
+	"gotest.tools/assert"
+)
+
+func TestSplitPackageName(t *testing.T) {
+	ns, name := splitPackageName("foo/bar")
+	assert.Equal(t, ns, "foo")
+	assert.Equal(t, name, "bar")
+
+	ns, name = splitPackageName("nonamespace")
+	assert.Equal(t, ns, "")
+	assert.Equal(t, name, "nonamespace")
+
+	ns, name = splitPackageName("some.repo.tk/v3/foo/bar")
+	assert.Equal(t, ns, "some.repo.tk/v3/foo")
+	assert.Equal(t, name, "bar")
+}
+
+var sampleMetadata = `
+name: machine
+namespace: heavy.metal
+version: "2000"
+maintainers:
+  - name: Billy Corgan
+    email: billy@pumpkins.net
+`
+
+var decodedSampleMetadata = types.AppMetadata{
+	Name:      "machine",
+	Namespace: "heavy.metal",
+	Version:   "2000",
+	Maintainers: []types.Maintainer{
+		{Name: "Billy Corgan", Email: "billy@pumpkins.net"},
+	},
+}
+
+func TestLoadMetadata(t *testing.T) {
```

Comment:
I don't know if this test is really usefull, it mostly tests the yaml package...

Reply:
So, my rules on those are :

- Test only exported function (or really complex non-exported functions), that way you don't "dig" into implementation detail (and your test is more future-proof)
- Don't re-test the library used. But I don't think that's the case here, i.e. it would be the case if we would test each and every case of yaml parsing (like errors, etc…). Here we *only* test that given the specified yaml, we get the specified struct — it's what the function does, whatever way it does, that's not what we test, we just test the "contract" here.

Reply:
PS: if the exported function are too complex to test — or take too much time — then it means it's doing too much on its own — or the abstraction level is not enough :angel:

Reply:
My point is more that is over testing is equally evil as under testing. It adds friction when refactoring, especially when testing un-exported functions. I think that 100% code coverage is not a healthy goal to achieve.

Reply:
This is a bit of a bizarre consideration to me, but if you want me to remove the tests I guess I can?

Reply:
nah, let's not remove the tests. My point is, *ideally* we should only test exported function — so in that case `Fork`. Go even has a magic trick for that, you can set the package of your test to `package_test` (here it would be `packager_test`) so that you only see exported one.

If it's too complicated to test, this mean the function is too complicated to test (too much responsibilities, too few abstraction, …). 

But as I said, this is *ideally* — won't block the PR for that :wink: